### PR TITLE
Parse TZif binary files

### LINKED ISSUE #1000: Parse TZif files into ICU4X data formats for TzdbDataProvider
One of the formats we should be able to generate ICU4X TimeZone data from is [TZif](https://tools.ietf.org/id/draft-murchison-tzdist-tzif-00.html).

We can use the [libtzfile](https://crates.io/crates/libtzfile) crate to parse this data into native Rust objects and then serialize the data to the ICU4X format of our choice. 

This will then match the combined `--tz-src-format tzif` and `--tz-src-path` in the `datagen` crate 

For more context see
* #1008

### PR DESCRIPTION
**Pull request by nordzilla**
Fixes #1000 

Draft PR for now. 

I need to take some time to split this into reviewable chunks.

### CODE REVIEW

**Comment by nordzilla** [comment]
## Description

This PR defines a parser for TZif binary format files with the intent to use them for a TZif data provider. 

TZif binary files can be generated from the [IANA database](https://www.iana.org/time-zones), and they are also provided on some operating systems e.g.  `/usr/share/zoneinfo` on Ubuntu.

This PR does not include the data-provider infrastructure to connect this data in ICU4X yet. There are still some unresolved questions that we need to discuss regarding how to properly map this data to time-zone IDs. 

For example, there is a TZif file for `Poland` that exists in the root directory of the IANA data (interestingly, not in `Europe`). But `Poland` is not a valid time-zone ID. `Europe/Warsaw` is a valid ID, but we need to determine how to programmatically associate the two, because the TZif files do not include any information about associated time-zone IDs. 

## Design Considerations

We had considered using the https://crates.io/crates/libtzfile crate which, at first, looked promising. It is a nice library, but upon further inspection I realized that it does not parse the POSIX time-zone strings contained within the TZif footers. This information is critical for generating future daylight savings time transitions, rather than only looking up historic transition dates. 

I decided to create a parser myself primarily for the reason above, but this approach also has other advantages:

* ICU4X owns can can maintain the code. 
* The parse infrastructure has no external dependencies outside of tools we already use (such as `eyre`).
* The parse infrastructure is generic and can be easily extended to parse the human-readable IANA files, or retrieve data from ICU files in the future. 
* This parser provides detailed error messages via `eyre`. 

> An example error message after intentionally corrupting the data:
> ```
> parsing Some("testdata/Poland")
> Error: parse_local_time_type_records(): failed to parse local time type records
> Caused by:
>    0: parse_exactly(): expected `6` items, but parsed `1` items
>    1: parse_local_time_type_record(): failed to parse local time type record
>    2: parse_is_dst(): failed to parse is_dst
>    3: parse_bool(): expected byte to be 0 or 1, but found `Some(4)`
> ```

## Reviewing

This is a very large PR (my apologies), but it would be difficult to separate this out into distinct PRs that make any sense without other contexts.

Each commit represents a cohesive section that should be easily reviewable in isolation. 

The general structure is as follows:
```
.
├── src
│  ├── core   // Core types and traits on which to build parsing functionality.
│  │  └── ext // Generic trait extensions providing additional parse capabilities.
│  ├── data   // Definitions of the data representations that will be parsed.
│  ├── parse  // Implementations of the parsers themselves.
│  └── reader // Infrastructure for reading parse data.
└── tests 
```

**Comment by nordzilla** [comment]
> @nordzilla from a quick glance, it seems like the provider and parser code are in the same crate. I think it would probably be useful for the TZif parser to be a separate crate so that it's more generally useable; with the provider doing a more narrow task.

Ah yes, in my original iteration I had this code in `utils/tzif_parse` or something similar.

But I moved it into the provider crate later. I'm happy to move it back out. 

In fact, since this code doesn't actually contain any of the provider infrastructure, it would be easiest to just change this crate name.

**Review by Manishearth — COMMENTED** [review]
still working through it but posting these comments now

File: experimental/tzif/src/parse/posix.rs [hunk 1]
```diff
@@ -0,0 +1,756 @@
...
+                b'<' => source.parse_arbitrary_zone_variant_name(),
+                _ => source.parse_alphabetic_zone_variant_name(),
+            })
+            .wrap_err("parse_zone_variant_name(): failed to parse zone variant name")
+    }
+
+    /// Parses an arbitrary time-zone variant name. This name must be enclosed in angled brackets, e.g. `<name>`.
+    fn parse_arbitrary_zone_variant_name(&mut self) -> ParseResult<String, Source> {
+        self.begin_chain()
+            .chain(|source| source.expect_byte(b'<'))
```
Comment (Manishearth) [inline experimental/tzif/src/parse/posix.rs:57]: issue: I'm not a huge fan of this tuple based parsing system: IME what most parser libraries do is sequentially parse things as separate statements and you don't need complex tuple impls to make that work

**Comment by Manishearth** [comment]
So, hmm, the PR looks _fine_, but I feel like we're spending a lot of time building a hand-rolled parser when Rust has tons of off-the-shelf parsers. Furthermore, it's unclear to me why so many things are trait based here, or why the parser guts are all public: if we want them to be public they should be a separate parsing utility crate IMO.

**Comment by nordzilla** [comment]
@Manishearth 

Thanks for all of the feedback!

Hmm, perhaps there are some misunderstandings here.

I am far from an expert on parsers, but I will try to explain my reasoning.

> (also is there a reason for not using a parsing crate like nom?)

In the past when I've experimented with parsing crates like `nom`, `lalrpop`, etc. I've found them to be convoluted, and I have always ended up abandoning them for something more simple, intuitive, and hand-rolled, much like this infrastructure. But I recognize that everyone has different opinions on what is "intuitive" or "simple". The preexisting parsing library that I've liked most is `combine`, but even that has some interesting "magic". 

> I feel like we're spending a lot of time building a hand-rolled parser when Rust has tons of off-the-shelf parsers.

Perhaps I misunderstood, but I thought our general consensus in ICU4X is that if there are reasonable alternatives to not pull in more third-party crates, then we should avoid doing so. ICU4X currently does not use any third-party parsing crates that I am aware of (based on grepping the toml files). 

I also somewhat disagree that this is necessarily a large investment. Analyzing the `core` directory of this file structure, there's about 1,000 lines of code:

```
===============================================================================
 Language            Files        Lines         Code     Comments       Blanks
===============================================================================
 Rust                   12         1090          882           36          172
 |- Markdown            12          178            0          174            4
 (Total)                           1268          882          210          176
===============================================================================
 Total                  12         1090          882           36          172
===============================================================================
```

I would consider that to be a pretty small investment. It's a generic but very targeted infrastructure that is capable of parsing the `TZif` files, including the POSIX string data contained within the footers. It would also be easily extendable to parse the human-readable IANA files directly, or the ICU data, which are the two other sources we've considered pulling data from. The design is such that it should be easy to pull data from any source (binary files, text files, strings, slices, etc.). 

I also think that the code it produces is extremely readable, with no macro magic involved:

```
    fn parse_dst_variant_info(
        &mut self,
        std_offset: Seconds,
    ) -> ParseResult<ZoneVariantInfo, Source> {
        self.begin_chain()
            .chain(PosixTzStringParser::parse_zone_variant_name)
            .chain(|source| {
                source
                    .maybe_parse(PosixTzStringParser::parse_offset_time)
                    .or_default_to(std_offset - Hours(1).as_seconds())
            })  
            .map_value(|(name, offset)| ZoneVariantInfo { name, offset })
            .wrap_err("parse_dst_variant_info(): failed to parse dst time-zone variant info")
    } 
```
> Furthermore, it's unclear to me why so many things are trait based here

The trait-based design is what makes all of the functionality for the parser easily extendable. In the future, if we need to add functionality for the convenience of parsing the human-readable IANA data, we would only need to add a trait and then include that trait in the raw-iana parser, and it wouldn't affect the way the other parsers work at all. 

I see trait-based designs as beneficial and composable approach that Rust makes easy, so I like to utilize it when possible. 

> or why the parser guts are all public

I suppose they don't have to be public. I just didn't see a reason for them not to be. This isn't functionality that will be exposed during the ICU4X runtime. It's just a tool to be used by ICU4X DataProvider during the data generation phase, which is not part of the primary ICU4X API, nor is it performance-critical code. The ICU4X data provider will just run this parser and extract data from it to transform it into ICU4X's preferred representation.

As I mentioned previously, my initial design was for this to be a sort of stand-alone utility that the data provider would include and then use. I'm happy to move it back there and make everything more restricted.

**Comment by Manishearth** [comment]
> I also somewhat disagree that this is necessarily a large investment. Analyzing the `core` directory of this file structure, there's about 1,000 lines of code:

A thing that quickly blows up the investment is that if you're designing a generic parser library you probably want there to be far more documentation and examples; which, if you'd like to do, I'm all for it, but it seems like a lot of work.

**Comment by nordzilla** [comment]
> and either I'd like that parsing library to be a separate, generic crate that we are committed to maintaining (or you can publish on your own github and maintain), or i'd like us to use a preexisting library. 

While I think the idea of creating and maintaining a parsing crate sounds like a fun project, I don't have the formal expertise studying parsers, the time, or the desire to want to create something external on my own GitHub that I might deem "generally useful" to people.

My goal in making this generic was not to make it appear to be a general-purpose parsing library. I only wanted to design something that could easily handle parsing data from all three data sources, so that substantial refactoring would not be needed when we move beyond TZif as a data source.

My goal was to reduce the burden of maintenance and dependencies, with the intent being that this would be a sort of inernal tool that is only used by ICU4X data provider, but perhaps this had the opposite effect.

**Comment by Manishearth** [comment]
> While I think the idea of creating and maintaining a parsing crate sounds like a fun project, I don't have the formal expertise studying parsers, the time, or the desire to want to create something external on my own GitHub that I might deem "generally useful" to people.

I think the current status of this PR _is_ that you've created a general purpose parser. I figured you would not want to spend effort on polishing it, which is why I'm suggesting we use an existing library. I said this mostly to demonstrate _why_ I think we ought to use an existing library. I don't like the current state of us having a somewhat-but-not-quite general parsing framework squirreled away in our timezones parser. Especially not one publicly exposed, but my issue still applies when it's private.

> My goal was to reduce the burden of maintenance and dependencies, with the intent being that this would be a sort of inernal tool that is only used by ICU4X data provider, but perhaps this had the opposite effect.

I'm coming at this from the angle of if we're building a parser for an established format, might as well make it a utils/ crate that we publish that others may use as well (though we can have low stability/maintenance commitments). I'm also coming at this from the angle where I'd have preferred if we used an external crate for the tzif parsing itself — but the existing ecosystem does not serve our use case well, so I'm fine with having our own parser, I just find it a stretch to be maintaining our own parser _along with_ a parsing framework.

I'd also like to hear what @zbraniecki and @sffc feel about this.

**Comment by Manishearth** [comment]
(to be clear, I don't want to completely block on this and ask you to rewrite, but I'd like to discuss this a bit more first)

**Comment by nordzilla** [comment]
> (to be clear, I don't want to completely block on this and ask you to rewrite, but I'd like to discuss this a bit more first)

Agreed, I'm certainly not against changing it to use a parsing library. I just thought that this would be the most desirable path. 

I learned a lot about how the TZif files and POSIX strings work, which is still great knowledge and a bulk of the work. 

I'm interested to hear from Shane and Zibi as well.

File: experimental/tzif/src/parse/posix.rs [hunk 2]
```diff
@@ -0,0 +1,768 @@
...
+    },
+};
+use eyre::Context;
+
+impl<T, Source> PosixTzStringParser<Source> for T
+where
+    Source: Clone + ParseInput<u8, Vec<u8>, Source>,
+    T: ParseInput<u8, Vec<u8>, Source>,
+{
+}
```
Comment (sffc) [inline experimental/tzif/src/parse/posix.rs:32]: Suggestion: An empty blanket impl like this suggests to me that you should instead do a wrapper struct.

```rust
pub struct PosixTzStringParser<Source: Clone + ParseInput<u8, Vec<u8>, Source>> {
    source: Source
}
```

And them impl all your methods on the struct.

I think there are several places where you should be able to do this.

File: experimental/tzif/src/core/mod.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+/// A trait describing a type that is able to provide parse input.
+/// `Item` is the type of item returned by [`ParseInput::next()`].
+/// `Collection` is the type of collection of multiple items returned by [`ParseInput::take()`]
+/// `Source` is a clonable source of parse input.
+pub trait ParseInput<Item, Collection, Source = Self>
+where
+    Source: Clone + ParseInput<Item, Collection, Source>,
+    Self: Sized,
```
Comment (sffc) [inline experimental/tzif/src/core/mod.rs:64]: Suggestion: Make the trait specific to what you need. Everywhere in this crate, you use the same type parameters:

```
ParseInput<u8, Vec<u8>, Source>
```

If this is all you need, then I would just make the ParseInput trait always assume those types, removing the type parameters.

**Review by sffc — COMMENTED** [review]
Praise: Very thorough and good documentation!

File: experimental/tzif/src/parse/posix.rs (hunk 2 above)
Comment (Manishearth) [reply experimental/tzif/src/parse/posix.rs:32]: Yeah this is what I meant when I said I wasn't clear on why some of these things needed to be traits.

**Comment by sffc** [comment]
Thanks @nordzilla for getting the [docs](http://icu4x-pr-artifacts.storage.googleapis.com/gha/1280eac2b99e915a39a53d59f94f93fec7e25f36/docs/icu_provider_tzif/index.html) uploaded!

I left some initial review comments. Regarding architecture, I think that it is reasonable for the tzif parsing crate to export the trait that it needs *for parsing tzif files*. However, it is harmful in my view to make traits that are *too* generic. I think the traits like `icu_provider_tzif::core::ParseInputCopy` are currently leaning on the side of too generic.

Basically, if I'm a consumer and I want to make a new input source for the tzif parser, then there should be a highly specific trait that I can implement to make my source plug in with the parser. I shouldn't have to reason about how to implement a highly complex trait with multiple generic arguments.

**Comment by nordzilla** [comment]
Closing in favor of https://github.com/unicode-org/icu4x/pull/2019
